lib/net/ssh/authentication/methods/publickey.rb

```diff
@@ -97,9 +97,11 @@ def authenticate_with(identity, next_service, username)
               pubkey_algorithms.each do |pk_alg|
                 case pk_alg
                 when "rsa-sha2-512", "rsa-sha2-256", "ssh-rsa"
-                  if authenticate_with_alg(identity, next_service, username, pk_alg, pk_alg)
-                    # success
-                    return true
+                  if session.transport.algorithms.supported_signature_algs.include? pk_alg
+                    if authenticate_with_alg(identity, next_service, username, type, pk_alg)
```

Comment:
I am a little concerned that this will cause connections that work today to fail, if EXT_INFO was not being negotiated correctly in the past due to compatibility issues, etc. But I don't know the spec well enough...